ngraph/test/backend/unary_test.hpp

```diff
@@ -0,0 +1,211 @@
+#include "util/test_case.hpp"
+#include "util/test_control.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+namespace
+{
+    // Tolerance can be specified in two forms (number of bits in mantissa or float-point
+    // threshold), corresponding to two different comparing methods implemented in TestCase
+    // classes. Following class acts as an encapsulation on these two options without introducing
+    // too many overloadings on the function requiring such argument.
+    struct ArgTolerance
+    {
+        const int tolerance_bits;
+        const float tolerance_fp;
+        bool using_bits;
+        ArgTolerance()
+            : tolerance_bits(-1)
+            , tolerance_fp(0)
+            , using_bits(true)
+        {
+        }
+        ArgTolerance(const int tolerance_bits)
+            : tolerance_bits(tolerance_bits)
+            , tolerance_fp(0)
+            , using_bits(true)
+        {
+        }
+        ArgTolerance(const float tolerance_fp)
+            : tolerance_bits(-1)
+            , tolerance_fp(tolerance_fp)
+            , using_bits(false)
+        {
+        }
+        ArgTolerance(const double tolerance_fp)
+            : tolerance_bits(-1)
+            , tolerance_fp(tolerance_fp)
+            , using_bits(false)
+        {
+        }
+    };
+
+    // ArgShape is used implicitly by caller of test_unary() in following ways:
+    //
+    //  {PartialShape{2,Dimension::dynamic(),3}, Shape{2, 4, 3}}
+    //        A full specified form with both partial shape for ngraph function
+    //        generation and a static shape for final input tensor.
+    //
+    //  Shape{2, 4, 3}
+    //        A pure static shape for both ngraph function generation and input tensor
+    //
+    //  {}
+    //        A unspecified shape, test_unary() will derive the actual shape from the size
+    //        of the input vector as 1D shape.
+    struct ArgShape
+    {
+        PartialShape dynamic_shape;
+        Shape static_shape;
+        bool initialzed;
+        ArgShape(const PartialShape& dynamic_shape, const Shape& static_shape)
+            : dynamic_shape(dynamic_shape)
+            , static_shape(static_shape)
+            , initialzed(true)
+        {
+        }
+        ArgShape(const Shape& static_shape)
+            : dynamic_shape(static_shape)
+            , static_shape(static_shape)
+            , initialzed(true)
+        {
+        }
+        ArgShape()
+            : dynamic_shape{}
+            , static_shape{}
+            , initialzed(false)
+        {
+        }
+    };
+
+    template <typename TestEngine,
+              element::Type_t element_type,
+              typename value_type = fundamental_type_for<element_type>>
+    void test_unary(std::shared_ptr<Function> f,
+                    const std::vector<value_type>& test_input,
+                    const std::vector<value_type>& test_expected,
+                    ArgShape ashape = {},
+                    ArgTolerance tol = {})
```

Comment:
IMHO we should use `NDArrayBase` here which contains shape and data.

Reply:
here the data may have much less number of elements than shape required, suppose we test 4D shape {2,3,4,5} but we only have 4 inputs&expected data. in this case test_unary can expand the input by repeat itself to generate those required 120 data

Reply:
Where don't check whole tensor?